### ISSUE TITLE
ECS Cluster Scaling

### ISSUE THREAD

**Issue by 5290charlie** [widdix/aws-cf-templates#112]
Hi guys! I love the recent updates to the ECS cluster templates and scaling based on "Scheduable" container counts.

I just encountered a strange state in which an ECS service (based on the `service-cluster-alb` template) was being updated at the same time the cluster ASG was scaling down. Somehow a task was started on an instance as it was being terminated, and the entire stack update process stalled out (update rollback is currently still stalled also).

I have tweaked the settings of the cluster stack so the scaling operations will be less frequent, but this is something I wanted to bring up. I wonder if there is a way to ensure the cluster instances are "drained" from the cluster before the ASG scaledown actually happens... Maybe this is an issue that can be brought to Amazon's attention as well.

Still trying to understand exactly what's happening, but thought I would open this issue to start a conversation about it. 

Let me know what you think!

**Comment by 5290charlie** [comment]
I see that most of this lifecycle management with the draining of cluster instances is already in the template :+1: ... I also see that the default lifecycle_action_result is `CONTINUE`. 

One thought is that maybe the default action could be `ABANDON` so if the lifecycle hook failed/timed out it wouldn't continue with the instance termination.

Another might be related to the timing of the delay time for all of the components (alarm, lifecyle, cooldown, etc)

**Comment by michaelwittig** [comment]
Hi Charlie,

I understand that you encountered an issue when performing a rolling update of an ECS service while the ECS cluster was scaling down.

From my understanding: If the cluster scales down, the ASG selects one EC2 instance to be terminated. Thie EC2 instance changes its ECS state to DRAINING. This means that no new containers are scheduled to this instance. Also, all existing containers are re-scheduled to other instances. As soon as no containers are running on the instance anymore, the instance deregisters from ECS and completes the ASG termination. Which results in the termination of the EC2 instance.

The ECS service should offer you a history of containers that were launched/stopped/etc. Could you paste this information so we may see what was going wrong?

Thanks
Michael

**Comment by 5290charlie** [comment]
Hey Michael,

I am unable to retrieve this history now, but everything has been stable since I updated the cluster scaling parameters. The main issue I saw when this happened was looking at the ECS service event history and it was attempting to "stop" a task, but when inspecting that task, it was stuck in a "RUNNING" state with a desired state of "STOPPED" and the EC2 instance it was "RUNNING" on was already terminated. Based on the configuration, the only way I can imagine this happening is the lifecycle hook to "DRAIN" the ECS cluster instance before terminating it from the ASG timed out, and the default result action is "CONTINUE". So maybe there is some timing that can be optimized to ensure the DRAINING completes before the ASG lifecyle result defaults to CONTINUE, or the default action could be ABANDON as to not terminate an instance that has not successfully completed DRAINING yet.

Sorry for the lack of concrete info, I'll see if I can come up with any more data that could help diagnose this issue. Let me know if you have any other thoughts.

Thanks,
Charlie
